src/effects/effects_base/internal/effectsprovider.cpp

```diff
@@ -191,206 +191,194 @@ muse::Ret EffectsProvider::performEffect(au3::Au3Project& project, Effect* effec
-                                      EffectSettingsAccess& access, std::function<void()> updateUI, bool dryOnly)
+muse::Ret EffectsProvider::doEffectPreview(EffectBase& effect, EffectSettings& settings, bool dryOnly)
 {
-    auto cleanup0 = effect.BeginPreview(access.Get());
-
-    // These are temporary state in the Effect object that are meant be moved to
-    // a new class EffectContext
-    const auto numTracks = effect.mNumTracks;
-    const auto rate = effect.mProjectRate;
-    const auto& factory = effect.mFactory;
-    auto& mT0 = effect.mT0;
-    auto& mT1 = effect.mT1;
-    auto& mTracks = effect.mTracks;
-    auto& mProgress = effect.mProgress;
-    auto& mIsPreview = effect.mIsPreview;
-
-    // Get certain immutable properties of the effect
-    const auto previewFullSelection = effect.PreviewsFullSelection();
-    const auto isLinearEffect = effect.IsLinearEffect();
-
-    if (numTracks == 0) { // nothing to preview
-        return;
+    //! ============================================================================
+    //! NOTE Step 1 - check conditions
+    //! ============================================================================
+    if (effect.mNumTracks == 0) {     // nothing to preview
+        return muse::make_ret(muse::Ret::Code::InternalError);
     }
 
     auto player = playback()->player();
     if (!player->isBusy()) {
         LOGW() << "can't play, maybe audio is busy";
-        return;
+        return muse::make_ret(muse::Ret::Code::InternalError);
     }
 
-    const auto FocusDialog = BasicUI::FindFocus();
-    assert(FocusDialog); // postcondition
+    //! ============================================================================
+    //! NOTE Step 2 - save origin context (state)
+    //! ============================================================================
+    struct EffectContxt {
+        double t0 = 0.0;
+        double t1 = 0.0;
+        std::shared_ptr<TrackList> tracks;
+        BasicUI::ProgressDialog* progress = nullptr;
+        bool isPreview = false;
+    };
+
+    const EffectContxt originCtx = { effect.mT0, effect.mT1, effect.mTracks, effect.mProgress, effect.mIsPreview };
+    auto restoreCtx = finally([&] {
+        effect.mT0 = originCtx.t0;
+        effect.mT1 = originCtx.t1;
+        effect.mTracks = originCtx.tracks;
+        effect.mProgress = originCtx.progress;
+        effect.mIsPreview = originCtx.isPreview;
+    });
+
+    // restore internal effect state
+    auto cleanup0 = effect.BeginPreview(settings);
 
-    double previewDuration;
-    bool isNyquist = effect.GetFamily() == NYQUISTEFFECTS_FAMILY;
-    bool isGenerator = effect.GetType() == EffectTypeGenerate;
+    //! ============================================================================
+    //! NOTE Step 3 - make new context (state)
+    //! ============================================================================
 
-    const double previewLen = effect.mT1 - effect.mT0;
+    EffectContxt newCtx;
```

Comment:
With aliases and with using class members directly, I am confused that they change in an arbitrary order and it is not clear what exactly is used where, something like that (schematically)
```
   someMethod(effect.mT1, ...)
   ...
   10 code lines
   ...
   effect.mT1 = someVal;
   ...
   10 code lines
   ...
   someMethod(effect.mT1, ...) <- It is not clear whether the original meaning or the modified one is used here 
```

with EffectContxt it look like
```
   someMethod(originCtx.t1, ...)
   ...
   10 code lines
   ...
   newCtx.t1 = someVal;
   ...
   10 code lines
   ...
   someMethod(newCtx.t1, ...) <- it is clear that the new calculated value is used here (whatever it may be)
```

I am also confused by the fact that the state of the effect (its fields) changes not in one place, but by little bit in different places... If some internal logic uses these fields, then it is difficult to keep track of when the original values ​​will be used, when the new values.

Reply:
It would be good to replace all these class members with a structure (context) and remove them from the effect, passing the value of this structure to methods as an argument.